levelmap.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2021 Siôn le Roux.  All rights reserved.
+// Use of this source code is subject to an MIT-style
+// licence which can be found in the LICENSE file.
+
+package main
+
+import (
+	"image"
+	"math"
+
+	"github.com/fzipp/astar"
+)
+
+type LevelMap [][]int
+
+
+func CreateMap(w, h int) LevelMap {
+	lmap := make([][]int, h)
+	for i := range lmap {
+		lmap[i] = make([]int, w)
+	}
+	
+	return lmap
+}
+
+// Neighbours implements the astar.Graph interface
+func (m LevelMap) Neighbours(p image.Point) []image.Point {
+	offsets := []image.Point{
+		image.Pt(0, -1), // North
+		image.Pt(1, 0),  // East
+		image.Pt(0, 1),  // South
+		image.Pt(-1, 0), // West
+	}
+	offsetsDiag := []image.Point{
+		image.Pt(1, -1),  // NorthEast
+		image.Pt(1, 1),   // SouthEast
+		image.Pt(-1, 1),  // SouthWest
+		image.Pt(-1, -1), // NorthWest
+	}
+	res := make([]image.Point, 0, 8)
+	
+	// Check avaialable diagonal neighbours (free only if there is a wide enough corridor)
+	for _, off := range offsetsDiag {
+		q := p.Add(off)
+		q1 := p.Add(image.Pt(0, off.Y))
+		q2 := p.Add(image.Pt(off.X, 0))
+		if m.isFreeAt(q) && m.isFreeAt(q1) && m.isFreeAt(q2) {
+			res = append(res, q)
+		}
+	}
+
+	// Check avaialable  neighbours
+	for _, off := range offsets {
+		q := p.Add(off)
+		if m.isFreeAt(q) {
+			res = append(res, q)
+		}
+	}
+	return res
+}
+
+// isFreeAt returns if the cell is free
+func (m LevelMap) isFreeAt(p image.Point) bool {
+	return m[p.Y][p.X] == 0
+}
+
+// distance calculates Euclidean distance between the points
+func distance(p, q image.Point) float64 {
```

Comment:
Hmmm this looks almost exactly the same as another utility function we have lying in main.go, the usage is even almost the same, I guess it's just too bad that the types differ, this one is using image.Point and the other is using float64. Oh well.

Reply:
Yes. We may also be able to use image.Point internally, instead of our own similar struct ??

Reply:
I don't recommend this unless you want to make use of lots of these Point-specific functions that are already implemented in the standard library, add, subtract, and checking whether it's within another rectangle etc. otherwise it's annoying. I did something similar in a previous game, I thought it would be so cute and clean if the whole game was modeled around the stdlib's image.Point and image.Rect etc. but then 50% of my code was just converting between `int` and `float64` 😭

Reply:
OK, thx